src/irr/video/COpenGLPipelineCache.h

```diff
@@ -0,0 +1,62 @@
+#ifndef __IRR_C_OPENGL_PIPELINE_CACHE_H_INCLUDED__
+#define __IRR_C_OPENGL_PIPELINE_CACHE_H_INCLUDED__
+
+#include "irr/video/IGPUPipelineCache.h"
+#include "COpenGLExtensionHandler.h"
+#include "irr/video/COpenGLSpecializedShader.h"
+#include "irr/video/COpenGLPipelineLayout.h"
+#include "irr/core/Types.h"
+#include <array>
+
+namespace irr { namespace video
+{
+
+class COpenGLPipelineCache final : public IGPUPipelineCache
+{
+public:
+	struct SCacheVal {
+		COpenGLSpecializedShader::SProgramBinary binary;
+		core::smart_refctd_ptr<COpenGLPipelineLayout> layout;
```

Comment:
shouldn't layout be the key? just like specialization would be?

Reply:
no because there's no way to make an operator< for it

Reply:
let's invent one (don't actually give it a `operator<` but use a `map` with a custom comparator functor struct.

You can treat a layout as a 4d vector/array of descriptor layouts, so implement a `<` operator like you would for a 4bit LSB int (first set `<`, then if first set `==` then check second set, and so on)

Then a `<` function for descriptor layouts could be defined the same as you define a `<` operator for an `std::array`

Reply:
its possible to implement a comparator using very similar logic to this:
https://github.com/buildaworldnet/IrrlichtBAW/pull/422/files#diff-73a7b1e23025a75063836a1f6c71770aR80

Reply:
so u're saying that having let's say layouts A, B,C and knowing relation of compatibility between them I can sort them in some order that will optimize process of searching the one compatible with some another layout D? I don't think so, how would it work? Compatibility between layouts A and B says nothing about compatibility A and D

Reply:
>its possible to implement a comparator using very similar logic to this:

u dont want just a comparator, u want a clearly defined relation

Reply:
> so u're saying that having let's say layouts A, B,C and knowing relation of compatibility between them I can sort them in some order that will optimize process of searching the one compatible with some another layout D? I don't think so, how would it work? 

Well in order for a SPIR-V to give you the same GLSL shader, the layout needs to be FULLY compatible, which means that all descriptor sets need to be identically defined.

Fun fact, how STL implements all algorithms in terms of just `<`:
A<B: `A<B`
A>B: `B<A`
A!=B: `A<B || B<A`
A==B: `!(A!=B)`

So we don't need a relation that makes sense, just one that is consistent (no `A<B<C && C<A` loop) and will guarantee that equality will work.

> Compatibility between layouts A and B says nothing about compatibility A and D
> u dont want just a comparator, u want a clearly defined relation

Well compatibility for a SHADER binding reordering requires full compatibility, so if A and B are compatible for shader X and C is compatible with A or B up to set 4, then C will be compatible as the descriptor set layouts are identically defined.

P.S. Another way to establish a `<` relation is to first compare shader hash, then specialization info, and then the computed (set,binding) to (type,binding) remappings which could be cached in `COpenGLPipelineLayout` similarily to precomputed multibind params 
https://github.com/Crisspl/IrrlichtBAW/blob/shaders/src/irr/video/COpenGLPipelineLayout.h#L73